pkg/controllers/subnetport/subnetport_controller.go

```diff
@@ -577,28 +601,136 @@ func (r *SubnetPortReconciler) addressBindingMapFunc(ctx context.Context, obj cl
+func (r *SubnetPortReconciler) collectAddressBindingGarbage(ctx context.Context, namespace, ipAddress *string) {
+	abList := &v1alpha1.AddressBindingList{}
+	var err error
+	if namespace == nil || *namespace == "" {
+		err = r.Client.List(context.TODO(), abList)
+	} else {
+		err = r.Client.List(context.TODO(), abList, client.InNamespace(*namespace))
+	}
+	if err != nil {
+		log.Error(err, "Failed to list AddressBinding from cache")
+		return
+	}
+	for i, ab := range abList.Items {
+		if ipAddress != nil && ab.Status.IPAddress != *ipAddress {
+			continue
+		}
+		spList := &v1alpha1.SubnetPortList{}
+		spIndexValue := fmt.Sprintf("%s/%s", ab.Namespace, ab.Spec.VMName)
+		err := r.Client.List(context.TODO(), spList, client.MatchingFields{util.SubnetPortNamespaceVMIndexKey: spIndexValue})
+		if err != nil {
+			log.Error(err, "Failed to list SubnetPort from cache", "indexValue", spIndexValue)
+			continue
+		}
+		if ab.Spec.InterfaceName == "" && len(spList.Items) > 1 {
+			setAddressBindingStatus(r.Client, ctx, &abList.Items[i], metav1.Now(), multipleInterfaceFoundError, "")
+			continue
+		}
+		found := false
+		for i, sp := range spList.Items {
+			vm, port, err := common.GetVirtualMachineNameForSubnetPort(&spList.Items[i])
+			if err != nil || vm == "" {
+				log.Error(err, "Failed to get VM name from SubnetPort", "namespace", sp.Namespace, "name", sp.Name, "annotations", sp.Annotations)
+				continue
+			}
+			if ab.Spec.InterfaceName == "" || ab.Spec.InterfaceName == port {
+				found = true
+				break
+			}
+		}
+		if !found {
+			setAddressBindingStatus(r.Client, ctx, &abList.Items[i], metav1.Now(), vmOrInterfaceNotFoundError, "")
+		}
+	}
+}
+
+func setAddressBindingStatusBySubnetPort(client client.Client, ctx context.Context, subnetPort *v1alpha1.SubnetPort, subnetPortService *subnetport.SubnetPortService, transitionTime metav1.Time, e error) {
+	ipAddress := ""
 	subnetPortID := subnetPortService.BuildSubnetPortId(&subnetPort.ObjectMeta)
 	nsxSubnetPort := subnetPortService.SubnetPortStore.GetByKey(subnetPortID)
 	if nsxSubnetPort == nil {
 		log.Info("Missing SubnetPort", "id", subnetPort.UID)
-		return
-	}
-	if nsxSubnetPort.ExternalAddressBinding == nil || nsxSubnetPort.ExternalAddressBinding.ExternalIpAddress == nil {
-		return
+		if e == nil {
+			e = fmt.Errorf("missing SubnetPort")
+		}
+	} else if nsxSubnetPort.ExternalAddressBinding != nil && nsxSubnetPort.ExternalAddressBinding.ExternalIpAddress != nil {
+		ipAddress = *nsxSubnetPort.ExternalAddressBinding.ExternalIpAddress
 	}
 	ab := subnetPortService.GetAddressBindingBySubnetPort(subnetPort)
 	if ab == nil {
-		log.Info("Missing AddressBinding for SubnetPort", "namespace", subnetPort.Namespace, "name", subnetPort.Name)
+		log.Info("No AddressBinding for SubnetPort", "namespace", subnetPort.Namespace, "name", subnetPort.Name)
```

Comment:
if there is ExternalAddressBinding in NSX subnetport, but there is no AddressBinding CR found, do we need reconcile again?

Reply:
No need to reconcile SubnetPort here since there will be an AddressBinding DELETE event for your case.